schema.json

```diff
@@ -1,255 +1,266 @@
+          "type": "integer",
+          "minimum": 0
+        },
+        "maxItems": {
+          "description": "The maximum list size.",
           "type": "integer",
-          "maximum": 4294967295,
           "minimum": 0
         }
       },
-      "required": ["label", "value"]
+      "required": [ "type", "itemType" ]
     },
-    "name": {
+    "nameType": {
+      "description": "A name is a string having a length of at least 1.",
       "type": "string",
       "minLength": 1
     },
-    "range": {
+    "oneOfTypes": {
+      "description": "One of any of the types. This provides a single location
+                      to keep the list.",
+      "oneOf": [
+        { "$ref": "#/definitions/bitFieldType" },
+        { "$ref": "#/definitions/compoundType" },
+        { "$ref": "#/definitions/integerType" },
+        { "$ref": "#/definitions/listType" },
+        { "$ref": "#/definitions/pdEnvelopeType" },
+        { "$ref": "#/definitions/simpleType" },
+        { "$ref": "#/definitions/stringType" },
+        { "$ref": "#/definitions/subdeviceType" }
+      ]
+    },
+    "subdeviceType": {
+      "description": "A subdevice type. One of the root device (e.g. value 0),
+                      the root or all subdevices or the broadcast subdevice,
+                      the root or any subdevice, or just any subdevice.",
+      "type" : "string",
+      "enum": [
+            "root",
+            "root_or_any_subdevice_or_broadcast",
+            "root_or_any_subdevice",
+            "any_subdevice"
+          ]
+    },
+    "pdEnvelopeType": {
+      "description": "Contains a length/data pair for one Parameter Data item,
+                      where the length is an unsigned 8-bit value and the data
+                      has 'length' bytes. This exists to provide a schema
+                      definition for the 'envelope' of a PDL/PD pair.",
+      "type": "object",
+      "properties": {
+        "type": { "const": "pdEnvelope" },
+        "length": {
+          "description": "The data length can be optionally specified.",
+          "type": "integer",
+          "minimum": 0
+        }
+      },
+      "required": [ "type" ]
+    },
+    "rangeType": {
+      "description": "Defines an inclusive range of numbers.",
       "type": "object",
       "properties": {
-        "lower": {
+        "minimum": {
+          "description": "The lower bound, inclusive.",
           "type": "integer"
         },
-        "upper": {
+        "maximum": {
+          "description": "The upper bound, inclusive.",
           "type": "integer"
         }
       },
-      "required": ["lower", "upper"]
+      "required": [ "minimum", "maximum" ]
+    },
+    "simpleType": {
+      "description": "Simple type that has no fields other than a name and
+                      type.",
+      "type": "object",
+      "properties": {
+        "name": { "$ref": "#/definitions/nameType" },
+        "type": {
+          "enum": [
```

Comment:
Should we add DS_NOT_DEFINED from Table A-15 in E1.20, I'm not sure what it's for and I don't think I've seen it in the wild.

Reply:
We also have no way to represent the Manufacturer-Specific Data Types in this schema, although I'm not sure what they'd be apart from the fabled uint128 and indeed how the console would use them.

Reply:
I added "int128" and "uint128". Can't hurt. Good suggestion.

Reply:
There's a good chance that any of the manufacturer data types could be represented using the existing types in the schema, especially since there's string, list, and compound types. "list-of-bytes" or strings could be used to represent pretty much all data.

Reply:
Agreed, although then you're back to the rather limited interface some controllers provide where manufacturer specific PID support is "please enter this list of bytes". For example if you were downloading a gobo preview, you really want to know what you're receiving is a BMP or a PNG or whatever, not just get sent a load of bytes.